server/src/main/java/io/deephaven/server/table/stats/ObjectChunkedStats.java

```diff
@@ -0,0 +1,117 @@
+
+import gnu.trove.map.hash.TObjectLongHashMap;
+import io.deephaven.chunk.ObjectChunk;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.datastructures.util.CollectionUtil;
+import io.deephaven.engine.rowset.RowSequence;
+import io.deephaven.engine.rowset.RowSet;
+import io.deephaven.engine.table.ChunkSource;
+import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.Table;
+import io.deephaven.engine.table.impl.util.ColumnHolder;
+import io.deephaven.engine.util.TableTools;
+import io.deephaven.util.QueryConstants;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+public class ObjectChunkedStats implements ChunkedStatsKernel {
+    private final int maxUniqueToCollect;
+    private final int maxUniqueToDisplay;
+
+    public ObjectChunkedStats(int maxUniqueToCollect, int maxUniqueToDisplay) {
+        this.maxUniqueToCollect = maxUniqueToCollect;
+        this.maxUniqueToDisplay = maxUniqueToDisplay;
+    }
+
+    @Override
+    public Table processChunks(final RowSet rowSet, final ColumnSource<?> columnSource, boolean usePrev) {
+        long count = 0;
+        int uniqueCount = 0;
+
+        final TObjectLongHashMap<Object> countValues = new TObjectLongHashMap<>();
+        boolean useSet = false;
+        final Set<Object> uniqueValues = new HashSet<>();
+
+        try (final ChunkSource.GetContext getContext = columnSource.makeGetContext(CHUNK_SIZE)) {
+            final RowSequence.Iterator rsIt = rowSet.getRowSequenceIterator();
+
+            while (rsIt.hasMore()) {
+                // Grab up to the next CHUNK_SIZE rows
+                final RowSequence nextKeys = rsIt.getNextRowSequenceWithLength(CHUNK_SIZE);
+
+                final ObjectChunk<?, ? extends Values> chunk =
+                        (usePrev ? columnSource.getPrevChunk(getContext, nextKeys)
+                                : columnSource.getChunk(getContext, nextKeys)).asObjectChunk();
+                final int chunkSize = chunk.size();
+                for (int ii = 0; ii < chunkSize; ii++) {
+                    final Object val = chunk.get(ii);
+
+                    if (val == null) {
+                        continue;
+                    }
+
+                    count++;
+
+                    if (useSet) {
+                        uniqueValues.add(val);
+                    } else if (uniqueCount > maxUniqueToCollect) {
+                        // we no longer need to track counts for these items; fall back to a Set to get at least a count
+                        uniqueValues.addAll(countValues.keySet());
+                        countValues.clear();
+                        uniqueValues.add(val);
+                        useSet = true;
+                    } else if (countValues.adjustOrPutValue(val, 1, 1) == 1) {
+                        uniqueCount++;
+                    }
```

Comment:
Seems to me like we do this pessimistically, before knowing if the new value increases our unique count. We can probably simplify:
```suggestion
                    } else if (countValues.adjustOrPutValue(val, 1, 1) == 1 && ++uniqueCount > maxUniqueToCollect) {
                        // we no longer need to track counts for these items; fall back to a Set to just get a count
                        uniqueValues.addAll(countValues.keySet());
                        countValues.clear();
                        useSet = true;
                    }
```

Reply:
what do you think about this instead, to avoid adding to one collection, copying to the other, clearing the first:

```
            while (iterator.hasNext()) {
                char val = iterator.next();
                if (val == QueryConstants.NULL_CHAR) {
                    continue;
                }
                count++;

                if (countValues.adjustOrPutValue(val, 1, 1) == 1 && ++uniqueCount > maxUniqueToCollect) {
                    // we no longer want to track counts for these items; fall back to a Set to get at least a count
                    uniqueValues.addAll(countValues.keySet());
                    countValues.clear();
                    useSet = true;
                    break;
                }
            }
            while (iterator.hasNext()) {
                char val = iterator.next();
                if (val == QueryConstants.NULL_CHAR) {
                    continue;
                }
                count++;

                uniqueValues.add(val);
            }
```

Reply:
This seems perfectly nice, but I don't understand what we're avoiding. You *seem* to have the same logic as my suggestion, just splitting into two iterations to avoid the `if` inside the iteration.